README.md
# Power Grid Lib - Unit Commitment

This benchmark library is curated and maintained by the [IEEE PES Task Force on Benchmarks for Validation of Emerging Power System Algorithms](https://power-grid-lib.github.io/) and is designed to evaluate a well established version of the the Unit Commitment problem.  Specifically, these cases are designed for benchmarking algorithms that solve the following the Mixed-Integer Linear Program described in the formulation [PDF document](MODEL.pdf).

A detailed description of this mathematical model is available [here](http://www.optimization-online.org/DB_FILE/2018/11/6930.pdf).  All of the cases files are curated in a json-based data format.  Open-source reference implementations are available in [EGRET](https://github.com/grid-parity-exchange/Egret) and [psst](https://github.com/kdheepak/psst).

## Problem Overview

These cases are useful for benchmarking solution methods for a variant of the unit commitment problem common in the academic literature. The features of this model are:
* A global load requirement with time series
* An optional global spinning reserve requirement with time series
* Thermal generators with technical parameters, including
    * Minimum and maximum power output
    * Hourly ramp-up and ramp-down rates
    * Start-up and shut-down ramp rates
    * Minimum run-times and off-times
    * Off time dependent start-up costs
    * Piecewise linear convex production costs
    * No-load costs
* Optional renewable generators with time series for minimum and maximum production.

## Case File Overview

This is divided into three sets of generators, each with varying load and reserve profiles. Please see the reference documents for more detail on the sources and motivations for these test cases.

### /ca/*.json
Generator and load data curated from publicly available data from the California ISO. There are four different load profiles by date, and varying reserve requirements as a percentage of load: 0%, 1%, 3%, and 5%. A fifth load profile (Scenario400) includes a hypothetical 40% wind generation supply. When using this data, please cite [1].

### /ferc/*.json
Generator data based on the publicly available unit commitment test instance from the Federal Energy Regulatory Commission. Load, reserves, and wind data is curated from publicly available data from PJM. The 'lw' moniker denotes a wind profile scaled to be 2% of annual load; the 'hw' denotes a wind profile scaled to be 30% of annual load. When using this data, please cite [1],[2].

### /rts_gmlc/*.json
Generator, load, and reserve data is curated from the publicly available [RTS-GMLC test system](https://github.com/GridMod/RTS-GMLC). Hourly generator ramp-rates are divided by a factor of three to ensure all the technical features listed above are captured at an hourly time scale. When using this data, please cite [3].

#### References
[1] Knueven, Bernard, James Ostrowski, and Jean-Paul Watson. "On mixed integer programming formulations for the unit commitment problem." Pre-print available at [http://www.optimization-online.org/DB_HTML/2018/11/6930.pdf](http://www.optimization-online.org/DB_HTML/2018/11/6930.pdf) (2018).

[2] Krall, Eric, Michael Higgins, and Richard P. O’Neill. "RTO unit commitment test system." Federal Energy Regulatory Commission. Available: [http://ferc.gov/legal/staff-reports/rto-COMMITMENT-TEST.pdf](http://ferc.gov/legal/staff-reports/rto-COMMITMENT-TEST.pdf) (2012).

[3] Barrows, Clayton, Aaron Bloom, Ali Ehlen, Jussi Ikaheimo, Jennie Jorgenson, Dheepak Krishnamurthy, Jessica Lau et al. "The IEEE Reliability Test System: A Proposed 2019 Update." IEEE Transactions on Power Systems (2019).


## Contributions

All case files are provided under a [Creative Commons Attribution License](http://creativecommons.org/licenses/by/4.0/), which allows anyone to share or adapt these cases as long as they give appropriate credit to the original author, provide a link to the license, and indicate if changes were made.

Community-based recommendations and contributions are welcome and encouraged in all PGLib repositories. Please feel free to submit comments and questions in the [issue tracker](https://github.com/power-grid-lib/pglib-uc/issues).  Corrections and new network contributions are welcome via pull requests.  All data contributions are subject to a quality assurance review by the repository curator(s).


## Citation Guidelines

This repository is not static.  Consequently, it is critically important to indicate the version number when referencing this repository in scholarly work.

Users of this these cases are encouraged to cite the original source documents mentioned in this overview document.


CHANGELOG.md
PGLib UC Benchmarks Change Log
==============================

### v19.08
- Initial release


uc_model.py
from pyomo.environ import *
import json
import sys

## Grab instance file from first command line argument
data_file = sys.argv[1]

print('loading data')
data = json.load(open(data_file, 'r'))

thermal_gens = data['thermal_generators']
renewable_gens = data['renewable_generators']

time_periods = {t+1 : t for t in range(data['time_periods'])}

gen_startup_categories = {g : list(range(0, len(gen['startup']))) for (g, gen) in thermal_gens.items()}
gen_pwl_points = {g : list(range(0, len(gen['piecewise_production']))) for (g, gen) in thermal_gens.items()}

print('building model')
m = ConcreteModel()

m.cg = Var(thermal_gens.keys(), time_periods.keys())
m.pg = Var(thermal_gens.keys(), time_periods.keys(), within=NonNegativeReals)  
m.rg = Var(thermal_gens.keys(), time_periods.keys(), within=NonNegativeReals)  
m.pw = Var(renewable_gens.keys(), time_periods.keys(), within=NonNegativeReals)
m.ug = Var(thermal_gens.keys(), time_periods.keys(), within=Binary) 
m.vg = Var(thermal_gens.keys(), time_periods.keys(), within=Binary) 
m.wg = Var(thermal_gens.keys(), time_periods.keys(), within=Binary) 

m.dg = Var(((g,s,t) for g in thermal_gens for s in gen_startup_categories[g] for t in time_periods), within=Binary) ##
m.lg = Var(((g,l,t) for g in thermal_gens for l in gen_pwl_points[g] for t in time_periods), within=UnitInterval) ##

m.obj = Objective(expr=sum(
                          sum(
                              m.cg[g,t] + gen['piecewise_production'][0]['cost']*m.ug[g,t]
                              + sum( gen_startup['cost']*m.dg[g,s,t] for (s, gen_startup) in enumerate(gen['startup']))
                          for t in time_periods)
                        for g, gen in thermal_gens.items() )
                        ) #(1)

m.demand = Constraint(time_periods.keys())
m.reserves = Constraint(time_periods.keys())
for t,t_idx in time_periods.items():
    m.demand[t] = sum( m.pg[g,t]+gen['power_output_minimum']*m.ug[g,t] for (g, gen) in thermal_gens.items() ) + sum( m.pw[w,t] for w in renewable_gens ) == data['demand'][t_idx] #(2)
    m.reserves[t] = sum( m.rg[g,t] for g in thermal_gens ) >= data['reserves'][t_idx] #(3)

m.uptimet0 = Constraint(thermal_gens.keys())
m.downtimet0 = Constraint(thermal_gens.keys())
m.logicalt0 = Constraint(thermal_gens.keys())
m.startupt0 = Constraint(thermal_gens.keys())

m.rampupt0 = Constraint(thermal_gens.keys())
m.rampdownt0 = Constraint(thermal_gens.keys())
m.shutdownt0 = Constraint(thermal_gens.keys())

for g, gen in thermal_gens.items():
    if gen['unit_on_t0'] == 1:
        if gen['time_up_minimum'] - gen['time_up_t0'] >= 1:
            m.uptimet0[g] = sum( (m.ug[g,t] - 1) for t in range(1, min(gen['time_up_minimum'] - gen['time_up_t0'], data['time_periods'])+1)) == 0 #(4)
    elif gen['unit_on_t0'] == 0:
        if gen['time_down_minimum'] - gen['time_down_t0'] >= 1:
            m.downtimet0[g] = sum( m.ug[g,t] for t in range(1, min(gen['time_down_minimum'] - gen['time_down_t0'], data['time_periods'])+1)) == 0 #(5)
    else:
        raise Exception('Invalid unit_on_t0 for generator {}, unit_on_t0={}'.format(g, gen['unit_on_t0']))

    m.logicalt0[g] = m.ug[g,1] - gen['unit_on_t0'] == m.vg[g,1] - m.wg[g,1] #(6)

    startup_expr = sum( 
                        sum( m.dg[g,s,t] 
                                for t in range(
                                                max(1,gen['startup'][s+1]['lag']-gen['time_down_t0']+1),
                                                min(gen['startup'][s+1]['lag']-1,data['time_periods'])+1
                                              )
                            ) 
                       for s,_ in enumerate(gen['startup'][:-1])) ## all but last
    if isinstance(startup_expr, int):
        pass
    else:
        m.startupt0[g] = startup_expr == 0 #(7)

    m.rampupt0[g] = m.pg[g,1] + m.rg[g,1] - gen['unit_on_t0']*(gen['power_output_t0']-gen['power_output_minimum']) <= gen['ramp_up_limit'] #(8)

    m.rampdownt0[g] = gen['unit_on_t0']*(gen['power_output_t0']-gen['power_output_minimum']) - m.pg[g,1] <= gen['ramp_down_limit'] #(9)


    shutdown_constr = gen['unit_on_t0']*(gen['power_output_t0']-gen['power_output_minimum']) <= gen['unit_on_t0']*(gen['power_output_maximum'] - gen['power_output_minimum']) - max((gen['power_output_maximum'] - gen['ramp_shutdown_limit']),0)*m.wg[g,1] #(10)

    if isinstance(shutdown_constr, bool):
        pass
    else:
        m.shutdownt0[g] = shutdown_constr

m.mustrun = Constraint(thermal_gens.keys(), time_periods.keys())
m.logical = Constraint(thermal_gens.keys(), time_periods.keys())
m.uptime = Constraint(thermal_gens.keys(), time_periods.keys())
m.downtime = Constraint(thermal_gens.keys(), time_periods.keys())
m.startup_select = Constraint(thermal_gens.keys(), time_periods.keys())
m.gen_limit1 = Constraint(thermal_gens.keys(), time_periods.keys())
m.gen_limit2 = Constraint(thermal_gens.keys(), time_periods.keys())
m.ramp_up = Constraint(thermal_gens.keys(), time_periods.keys())
m.ramp_down = Constraint(thermal_gens.keys(), time_periods.keys())
m.power_select = Constraint(thermal_gens.keys(), time_periods.keys())
m.cost_select = Constraint(thermal_gens.keys(), time_periods.keys())
m.on_select = Constraint(thermal_gens.keys(), time_periods.keys())

for g, gen in thermal_gens.items():
    for t in time_periods:
        m.mustrun[g,t] = m.ug[g,t] >= gen['must_run'] #(11)

        if t > 1:
            m.logical[g,t] = m.ug[g,t] - m.ug[g,t-1] == m.vg[g,t] - m.wg[g,t] #(12)

        UT = min(gen['time_up_minimum'],data['time_periods'])
        if t >= UT:
            m.uptime[g,t] = sum(m.vg[g,t] for t in range(t-UT+1, t+1)) <= m.ug[g,t] #(13)
        DT = min(gen['time_down_minimum'],data['time_periods'])
        if t >= DT:
            m.downtime[g,t] = sum(m.wg[g,t] for t in range(t-DT+1, t+1)) <= 1-m.ug[g,t] #(14)
        m.startup_select[g,t] = m.vg[g,t] == sum(m.dg[g,s,t] for s,_ in enumerate(gen['startup'])) #(16)

        m.gen_limit1[g,t] = m.pg[g,t]+m.rg[g,t] <= (gen['power_output_maximum'] - gen['power_output_minimum'])*m.ug[g,t] - max((gen['power_output_maximum'] - gen['ramp_startup_limit']),0)*m.vg[g,t] #(17)

        if t < len(time_periods): 
            m.gen_limit2[g,t] = m.pg[g,t]+m.rg[g,t] <= (gen['power_output_maximum'] - gen['power_output_minimum'])*m.ug[g,t] - max((gen['power_output_maximum'] - gen['ramp_shutdown_limit']),0)*m.wg[g,t+1] #(18)

        if t > 1:
            m.ramp_up[g,t] = m.pg[g,t]+m.rg[g,t] - m.pg[g,t-1] <= gen['ramp_up_limit'] #(19)
            m.ramp_down[g,t] = m.pg[g,t-1] - m.pg[g,t] <= gen['ramp_down_limit'] #(20

        piece_mw1 = gen['piecewise_production'][0]['mw']
        piece_cost1 = gen['piecewise_production'][0]['cost']
        m.power_select[g,t] = m.pg[g,t] == sum( (piece['mw'] - piece_mw1)*m.lg[g,l,t] for l,piece in enumerate(gen['piecewise_production'])) #(21)
        m.cost_select[g,t] = m.cg[g,t] == sum( (piece['cost'] - piece_cost1)*m.lg[g,l,t] for l,piece in enumerate(gen['piecewise_production'])) #(22)
        m.on_select[g,t] = m.ug[g,t] == sum(m.lg[g,l,t] for l,_ in enumerate(gen['piecewise_production'])) #(23)

m.startup_allowed = Constraint(m.dg_index)
for g, gen in thermal_gens.items():
    for s,_ in enumerate(gen['startup'][:-1]): ## all but last
        for t in time_periods:
            if t >= gen['startup'][s+1]['lag']:
                m.startup_allowed[g,s,t] = m.dg[g,s,t] <= sum(m.wg[g,t-i] for i in range(gen['startup'][s]['lag'], gen['startup'][s+1]['lag'])) #(15)

for w, gen in renewable_gens.items():
    for t, t_idx in time_periods.items():
        m.pw[w,t].setlb(gen['power_output_minimum'][t_idx]) #(24)
        m.pw[w,t].setub(gen['power_output_maximum'][t_idx]) #(24)

print("model setup complete")

from pyomo.opt import SolverFactory
cbc = SolverFactory('cbc')

print("solving")
cbc.solve(m, options={'ratioGap':0.01}, tee=True)


uc_model.jl
# Tested with Julia v1.1, JSON v0.21, JuMP v0.19, Cbc v0.6

using JSON
using JuMP


println("loading data")

## Grab instance file from first command line argument
data_file = ARGS[1]
data = JSON.parsefile(data_file)


println("model building")

thermal_gens = keys(data["thermal_generators"])
renewable_gens = keys(data["renewable_generators"])
time_periods = 1:data["time_periods"]

gen_startup_categories = Dict(g => 1:length(gen["startup"]) for (g,gen) in data["thermal_generators"])
gen_pwl_points = Dict(g => 1:length(gen["piecewise_production"]) for (g,gen) in data["thermal_generators"])

m = Model()

@variable(m, cg[thermal_gens,time_periods])
@variable(m, pg[thermal_gens,time_periods] >= 0)
@variable(m, pw[renewable_gens,time_periods] >= 0)
@variable(m, rg[thermal_gens,time_periods] >= 0)
@variable(m, ug[thermal_gens,time_periods], binary=true)
@variable(m, vg[thermal_gens,time_periods], binary=true)
@variable(m, wg[thermal_gens,time_periods], binary=true)
@variable(m, delta_sg[g in thermal_gens,gen_startup_categories[g],time_periods], binary=true)
@variable(m, 0 <= lambda_lg[g in thermal_gens,gen_pwl_points[g],time_periods] <= 1)


@objective(m, Min,
    sum(
        sum(
            cg[g,t] + gen["piecewise_production"][1]["cost"]*ug[g,t] +
            sum(
                gen_startup["cost"]*delta_sg[g,i,t]
            for (i, gen_startup) in enumerate(gen["startup"]))
        for t in time_periods)
    for (g, gen) in data["thermal_generators"])
) # (1)

for (g, gen) in data["thermal_generators"]

    if gen["unit_on_t0"] == 1
        @constraint(m, sum( (ug[g,t]-1) for t in 1:min(data["time_periods"], gen["time_up_minimum"] - gen["time_up_t0"]) ) == 0) # (4)
    else
        @constraint(m, sum( ug[g,t] for t in 1:min(data["time_periods"], gen["time_down_minimum"] - gen["time_down_t0"]) ) == 0) # (5)
    end

    @constraint(m, ug[g,1] - gen["unit_on_t0"] == vg[g,1] - wg[g,1]) # (6)

    @constraint(m, 0 ==
        sum(
            sum(
                delta_sg[g,i,t]
            for t in max(1, gen["startup"][i+1]["lag"] - gen["time_down_t0"] + 1):min(gen["startup"][i+1]["lag"]-1, data["time_periods"]))
        for (i,startup) in enumerate(gen["startup"][1:end-1]))
    ) # (7)

    @constraint(m, pg[g,1] + rg[g,1] - gen["unit_on_t0"]*(gen["power_output_t0"] - gen["power_output_minimum"]) <= gen["ramp_up_limit"]) # (8)
    @constraint(m, gen["unit_on_t0"]*(gen["power_output_t0"] - gen["power_output_minimum"]) - pg[g,1] <= gen["ramp_down_limit"]) # (9)
    @constraint(m, gen["unit_on_t0"]*(gen["power_output_t0"] - gen["power_output_minimum"]) <= gen["unit_on_t0"]*(gen["power_output_maximum"] - gen["power_output_minimum"]) - max(0, gen["power_output_maximum"] - gen["ramp_shutdown_limit"])*wg[g,1]) # (10)
end

for t in time_periods
    @constraint(m, 
        sum( pg[g,t] + gen["power_output_minimum"]*ug[g,t] for (g, gen) in data["thermal_generators"] ) +
        sum( pw[g,t] for g in renewable_gens)
        == data["demand"][t]
    ) # (2)

    @constraint(m, sum(rg[g,t] for g in thermal_gens) >= data["reserves"][t]) # (3)

    for (g, gen) in data["thermal_generators"]

        @constraint(m, ug[g,t] >= gen["must_run"]) # (11)

        if t > 1
            @constraint(m, ug[g,t] - ug[g,t-1] == vg[g,t] - wg[g,t]) # (12)
            @constraint(m, pg[g,t] + rg[g,t] - pg[g,t-1] <= gen["ramp_up_limit"]) # (19)
            @constraint(m, pg[g,t-1] - pg[g,t] <= gen["ramp_down_limit"]) # (20)
        end


        if t >= gen["time_up_minimum"] || t == data["time_periods"]
            @constraint(m, sum( vg[g,t2] for t2 in (t-min(gen["time_up_minimum"],data["time_periods"])+1):t) <= ug[g,t])  # (13)
        end

        if t >= gen["time_down_minimum"] || t == data["time_periods"]
            @constraint(m, sum( wg[g,t2] for t2 in (t-min(gen["time_down_minimum"],data["time_periods"])+1):t) <= 1 - ug[g,t])  # (14)
        end

        for (si,startup) in enumerate(gen["startup"][1:end-1])
            if t >= gen["startup"][si+1]["lag"]
                time_range = startup["lag"]:(gen["startup"][si+1]["lag"]-1)
                @constraint(m, delta_sg[g,si,t] <= sum(wg[g,t-i] for i in time_range)) # (15)
            end
        end

        @constraint(m, vg[g,t] == sum( delta_sg[g,i,t] for i in gen_startup_categories[g])) # (16)

        @constraint(m, pg[g,t] + rg[g,t] <= (gen["power_output_maximum"] - gen["power_output_minimum"])*ug[g,t] - max(0, (gen["power_output_maximum"] - gen["ramp_startup_limit"]))*vg[g,t]) # (17)

        if t < data["time_periods"]
            @constraint(m, pg[g,t] + rg[g,t] <= (gen["power_output_maximum"] - gen["power_output_minimum"])*ug[g,t] - max(0, (gen["power_output_maximum"] - gen["ramp_shutdown_limit"]))*wg[g,t+1]) # (18)
        end

        @constraint(m, pg[g,t] == sum((gen["piecewise_production"][l]["mw"] - gen["piecewise_production"][1]["mw"])*lambda_lg[g,l,t] for l in gen_pwl_points[g])) # (21)
        @constraint(m, cg[g,t] == sum((gen["piecewise_production"][l]["cost"] - gen["piecewise_production"][1]["cost"])*lambda_lg[g,l,t] for l in gen_pwl_points[g])) # (22)
        @constraint(m, ug[g,t] == sum(lambda_lg[g,l,t] for l in gen_pwl_points[g])) # (23)
    end

    for (rg, rgen) in data["renewable_generators"]
        @constraint(m, rgen["power_output_minimum"][t] <= pw[rg,t] <= rgen["power_output_maximum"][t]) # (24)
    end
end


println("optimization")

using Cbc
optimize!(m, with_optimizer(Cbc.Optimizer, logLevel=1))
